pydrive2/storage.py

```diff
@@ -0,0 +1,117 @@
+import os
+import json
+import warnings
+from filelock import FileLock
+
+
+_SYM_LINK_MESSAGE = "File: {0}: Is a symbolic link."
+_IS_DIR_MESSAGE = "{0}: Is a directory"
+_MISSING_FILE_MESSAGE = "Cannot access {0}: No such file or directory"
+
+
+def validate_file(filename):
+    if os.path.islink(filename):
+        raise IOError(_SYM_LINK_MESSAGE.format(filename))
+    elif os.path.isdir(filename):
+        raise IOError(_IS_DIR_MESSAGE.format(filename))
+    elif not os.path.isfile(filename):
+        warnings.warn(_MISSING_FILE_MESSAGE.format(filename))
+
+
+class CredentialBackend(object):
+    """Adapter that provides a consistent interface to read and write credential files"""
+
+    def _read_credentials(self, rpath):
+        """Specific implementation of how the storage object should retrieve a file."""
+        return NotImplementedError
+
+    def _store_credentials(self, credential, rpath):
+        """Specific implementation of how the storage object should write a file"""
+        return NotImplementedError
+
+    def _delete_credentials(self, rpath):
+        """Specific implementation of how the storage object should delete a file."""
+        return NotImplementedError
+
+    def read_credentials(self, rpath):
+        """Reads a credential config file and returns the config as a dictionary
+        :param fname: host name of the local web server.
+        :type host_name: str.`
+        :return: A credential file
+        """
+        return self._read_credentials(rpath)
+
+    def store_credentials(self, credential, rpath):
+        """Write a credential to
+        The Storage lock must be held when this is called.
+        Args:
+            credentials: Credentials, the credentials to store.
+        """
+        self._store_credentials(credential, rpath)
+
+    def delete_credentials(self, rpath):
+        """Delete credential.
+        Frees any resources associated with storing the credential.
+        The Storage lock must *not* be held when this is called.
+
+        Returns:
+            None
+        """
+        self._delete_credentials(rpath)
+
+
+class FileBackend(CredentialBackend):
+    # https://stackoverflow.com/questions/37084682/is-oauth-thread-safe
+    """Read and write credentials to a file backend with File Locking"""
+
+    def __init__(self):
+        self._locks = {}
+
+    def createLock(self, rpath):
+        self._locks[rpath] = FileLock("{}.lock".format(rpath))
```

Comment:
The previous implementation was Semaphore / Mutex or something like this-based. File based locks are problematic. Primarily because of something happens you would have to go and delete them manually. Let's please stick for now with what we had in the code before. For multi-threaded safety it's enough to use an in-memory lock.

When we write into a file (new credentials), we can guarantee multi-process safety with writing into a temp file first and then doing `mv`.

Reply:
I'm slightly confused by the purpose of any sort of thread locks for the Credentials object. I'm under the impression that there should only be one GoogleAuth object with `self.credentials` shared across all threads & the credentials would be refreshed synchronously. 

Some sources that suggest that credentials should be synchronously refreshed:
- https://stackoverflow.com/questions/37084682/is-oauth-thread-safe
- https://developers.google.com/adwords/api/docs/guides/authentication#credential_sharing_strategies

Reply:
I will update this to the `create temp file` -> `os.replace(temp_file, actual_filel)` implementation instead of filelock

Reply:
> the credentials would be refreshed synchronously.

I think that's why those locks exist, since in the current implementation multiple threads could try to refresh (and write it back to a file) them simultaneously. With the current implementation I'm not sure we are writing the credentials on refresh. It would be another difference in the implementation that we need to double check and see if there are some important implications.

> I will update this to the create temp file -> os.replace(temp_file, actual_filel) implementation instead of filelock

thanks.